packages/core/src/modules/routing/RecipientModule.ts

```diff
@@ -71,6 +73,57 @@ export class RecipientModule {
     }
   }
 
+  private async openMediationWebSocket(mediator: MediationRecord) {
+    const { message, connectionRecord } = await this.connectionService.createTrustPing(mediator.connectionId)
+
+    const websocketSchemes = ['ws', 'wss']
+    const hasWebSocketTransport =
+      connectionRecord.didDoc.didCommServices.filter((s) => websocketSchemes.includes(s.protocolScheme)).length > 0
+
+    if (!hasWebSocketTransport) {
+      throw new AriesFrameworkError('Cannot open websocket to connection without websocket service endpoint')
+    }
+
+    await this.messageSender.sendMessage(createOutboundMessage(connectionRecord, message), {
+      transportPriority: {
+        schemes: websocketSchemes,
+        restrictive: true,
+        // TODO: add keepAlive: true to enforce through the public api
+        // we need to keep the socket alive. It already works this way, but would
+        // be good to make more explicit from the public facing API.
+        // This would also make it easier to change the internal API later on.
+        // keepAlive: true,
+      },
+    })
+  }
+
+  private async initiateImplicitPickup(mediator: MediationRecord) {
+    let interval = 50
+
+    // Listens to Outbound websocket closed events and will reopen the websocket connection
+    // in a recursive back off strategy if it matches the following criteria:
+    // - Agent is not shutdown
+    // - Socket was for current mediator connection id
+    this.eventEmitter
+      .observable<OutboundWebSocketClosedEvent>(TransportEventTypes.OutboundWebSocketClosedEvent)
+      .pipe(
+        // Stop when the agent shuts down
+        takeUntil(this.agentConfig.stop$),
+        filter((e) => e.payload.connectionId === mediator.connectionId),
+        // Make sure we're not reconnecting multiple times
+        throttleTime(interval),
+        // Increase the interval (recursive back-off)
+        tap(() => (interval *= 2)),
+        // Wait for interval time before reconnecting
+        delay(interval)
+      )
+      .subscribe(() => {
+        this.openMediationWebSocket(mediator)
+      })
+
+    await this.openMediationWebSocket(mediator)
+  }
```

Comment:
Shouldn't those methods be part of the transport layer rather than inside the core of the framework? Or, it's just temporary?

Reply:
What it does is sending a trust ping message with a restriction to only use websocket. The websocket will be kept alive autmoatically (I'd like to make the keep alive more explicit)

I'm not sure yet how we could nicely integrate this into the transport layer. Would be nice to not make it WebSocket specific (so just send a ping and keep the socket alive), but we need to specify somewhere which transports are capable of this, as e.g. HTTP isn't. 

Maybe a transport should be able to specify whether it can send multiple messages (WebSocket is ∞, while HTTP is 1 request, 1 response). That way we wouldn't have to specify WS explicitly, but rather whether we want a transport that can stay open (for implicit pickup)

I'm going to think about it for a while. We can have a discussion about it in the WG call

Reply:
I have opened https://github.com/hyperledger/aries-framework-javascript/discussions/435. Please leave your thoughts :)